qiskit_experiments/library/randomized_benchmarking/rb_experiment.py

```diff
@@ -133,86 +142,128 @@ def circuits(self) -> List[QuantumCircuit]:
-                group_elt_gate = group_elt_circ[0]
-                group_elt_op = group_elt_circ[1]
-            else:
-                group_elt_gate = group_elt_circ
-                group_elt_op = Clifford(group_elt_circ)
-
-            if not isinstance(group_elt_gate, Instruction):
-                group_elt_gate = group_elt_gate.to_instruction()
-            circ_op = circ_op.compose(group_elt_op)
-            for circ in circs:
-                circ.append(group_elt_gate, qubits)
+            for elem in seq:
+                circ.append(self._to_instruction(elem), qubits)
                 circ.barrier(qubits)
-            if current_length + 1 in lengths:
-                # copy circuit and add inverse
-                inv = circ_op.adjoint()
-                rb_circ = circs.pop()
-                rb_circ.append(inv, qubits)
-                rb_circ.barrier(qubits)
-                rb_circ.metadata = {
-                    "experiment_type": self._type,
-                    "xval": current_length + 1,
-                    "group": "Clifford",
-                    "physical_qubits": self.physical_qubits,
-                }
-                rb_circ.measure_all()
-                circuits.append(rb_circ)
+
+            # Compute inverse, compute only the difference from the previous shorter sequence
+            for elem in seq[len(prev_seq) :]:
+                prev_elem = self._clifford_compose(prev_elem, elem)
+            prev_seq = seq
+            inv = self._clifford_adjoint(prev_elem)
+
+            circ.append(self._to_instruction(inv), qubits)
+            circ.barrier(qubits)  # TODO: Can we remove this? (measure_all inserts one more barrier)
+            circ.measure_all()
+            circuits.append(circ)
         return circuits
 
+    def _sample_sequence(self, length: int, rng: Generator) -> Sequence[SequenceElementType]:
+        """Sample a RB sequence with the given length.
+
+        Args:
+            length: RB sequences length.
+            rng: Generator object for random number generation.
+
+        Returns:
+            A RB sequence.
+        """
+        # Return integer instead of Clifford object for 1 or 2 qubit case
+        if self.num_qubits == 1:
+            return rng.integers(24, size=length)
+        if self.num_qubits == 2:
+            return rng.integers(11520, size=length)
+
+        return [random_clifford(self.num_qubits, rng) for _ in range(length)]
+
+    def _to_instruction(self, elem: SequenceElementType) -> Instruction:
+        # TODO: basis transformation in 1Q (and 2Q) cases for speed
+        # Switching for speed up
+        if isinstance(elem, Integral):
+            if self.num_qubits == 1:
+                return _clifford_1q_int_to_instruction(elem)
+            if self.num_qubits == 2:
+                return _clifford_2q_int_to_instruction(elem)
+        return elem.to_instruction()
+
+    def _identity_clifford(self) -> SequenceElementType:
```

Comment:
I suggest moving all the integer-based Clifford encoding methods to `clifford_utils.py` so the user of the experiment is left exposed to the implementation details of the optimization as little as possible and the developers have all the optimization-related code in one place

Reply:
This apply also to `_clifford_compose` and `_clifford_adjoint`

Reply:
Thanks, that's a good point we need to think of.
IMO, `clifford_utils` should be useful for optimizing the implementation of `StarndardRB` but it also should be independent of the optimization strategy of `StarndardRB` (i.e. using integers for 1- and 2-qubit Cliffords) as much as possible. For example, we might end up with using integers only for 1-qubit Cliffords.
So I intentionally keep the minimal codes to tell the optimization strategy here while putting the details of the utility functions used for optimization to `clifford_utils` (e.g. see `_to_instruction`). I agree that we should not include all of the implementation details here in StarndardRB but I think we should also need to let the StarndardRB developers know how the code is optimized within StarndardRB. (For example, I think we should have clifford_1q_compose or something function/table in `clifford_utils` but we should call it in  `StarndardRB`). If we put all the details including the optimization strategy to `clifford_utils`, the developers must read and maintain both `StarndardRB` and `clifford_utils` to understand and update the optimization strategy (that would decrease the readability and maintenability).

Reply:
I agree with Eli's comment above regarding `CliffordUtils`. In my implementation, in PR #892 , I actually moved more functionality to `CliffordUtils`. I also added to class parameters to its init - `num_qubits` and `basis_gates`. I thought it looked cleaner that way, because `StandardRB` now creates a `CliffordUtils` object once, and `CliffordUtils` then takes care of all the details, such as creating the transpiled circuits.

Reply:
Thanks Itoko, I understand. My hope though was that the flow of the experiment itself, `StandardRB` in this case, would not have to change even if we change the implementation details of the optimization. So if there is a way to have the experiment code work with Cliffords (e.g. random sampling, composing, inverse) in a somewhat abstract way, without caring about this or that implementation detail of the optimization, it would be much cleaner

Reply:
OK, I see I can try to put a bit more logics (around integer Clifford operations) into `clifford_utils`. However, I'd like to push such a refactoring of `clifford_utils` into a follow-up PR. I now understand this PR should focus on the refactoring of StandardRB/InterleavedRB. To make that clear, I'll remove the API changes in `clifford_utils` from this PR (i.e. separate it out to a follow-up PR). Does it make sense?
In the follow-up PR, pushing Merav's idea further, I'd like to discuss a new class something like `IndexedClifford(num_qubits, index)` that implements `adjoint` and `compose` methods like the regular `Clifford`. After this PR, such a further refactoring would become much easier. (e.g. we can remove _identity_clifford, _clifford_adjoint and _clifford_compose whenever we want since they are all private methods.)

Reply:
Reverted refactoring with API changes in clifford_utils in [6759eaf](https://github.com/Qiskit/qiskit-experiments/pull/898/commits/6759eafdff2ffeabec16db47690ff74d6f02c8b5)